### PR TITLE
[XRay] Add `-fxray-default-options` to pass build-time defined XRay options

### PR DESCRIPTION
**Pull request by mshockwave**
This flag specifies XRay options that will automatically be applied to the instrumented binaries during run-time even without setting `XRAY_OPTIONS`. This is useful in cases where setting the `XRAY_OPTIONS` environment variable might be difficult. Plus, it's a convenient way to populate XRay options when you always want the instrumentation to be enabled.

The implementation follows what memory profiler and PGO did for their profile file name and profile version, respectively: insert a weak linkage global variable carrying the option string, which is then intercepted by compiler-rt during xray's option parsing phase.

### CODE REVIEW

**Comment by mshockwave** [comment]
ping

File: compiler-rt/lib/xray/xray_options_var.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*===----- xray_options_var.cpp - XRay option variable setup  -------------===*\
```
Comment (MaskRay) [inline compiler-rt/lib/xray/xray_options_var.cpp:1]: `//`

File: compiler-rt/lib/xray/xray_flags.h [hunk 2]
```diff
@@ -17,6 +17,12 @@
 #include "sanitizer_common/sanitizer_flag_parser.h"
 #include "sanitizer_common/sanitizer_internal_defs.h"
 
+#define XRAY_OPTIONS_VAR __llvm_xray_options
```
Comment (MaskRay) [inline compiler-rt/lib/xray/xray_flags.h:20]: It doesn't seem useful to introduce a capitalized alias `XRAY_OPTIONS_VAR`. 

If a TU doesn't include `xray_flags.h`, the compiled code will reference `XRAY_OPTIONS_VAR`.

File: compiler-rt/test/xray/TestCases/Posix/default-options.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_xray -g -fxray-default-options='patch_premain=true:verbosity=1:xray_mode=xray-basic' -o %t %s
```
Comment (MaskRay) [inline compiler-rt/test/xray/TestCases/Posix/default-options.cpp:1]: lld/test/ELF style `// RUN: rm -fr %t && mkdir %t && cd %t` seems better

**Comment by MaskRay** [comment]
Perhaps `__xray_default_options()` like `__asan_default_options()`? The user would need to provide a `.o` that defines the function, but there is more flexibility (e.g. the function can inspect a config file).

**Comment by mshockwave** [comment]
> Perhaps `__xray_default_options()` like `__asan_default_options()`? The user would need to provide a `.o` that defines the function, but there is more flexibility (e.g. the function can inspect a config file).

I think this will be a better idea. It simplifies the design and no longer requires changes in Clang driver and the additional LLVM Pass (i.e. XRayPreparation). I implemented that in #117921 and will close this PR in favor of that one.

**Comment by mshockwave** [comment]
Close this PR in favor of #117921

**Pull request closed without merging**